tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardSetMethod.scala

```diff
@@ -0,0 +1,138 @@
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  override def doProcess(capabilities: Set[CapabilityIdentifier], invocation: InvocationWithContext,
+                         mailboxSession: MailboxSession,
+                         request: ForwardSetRequest): Publisher[InvocationWithContext] =
+    updateForward(request, mailboxSession)
+      .map(updateResult => createResponse(invocation.invocation, request, updateResult))
+      .map(InvocationWithContext(_, invocation.processingContext))
+
+  private def updateForward(request: ForwardSetRequest, mailboxSession: MailboxSession): SMono[ForwardSetUpdateResults] =
+    SFlux.fromIterable(request.parseUpdateRequest())
+      .flatMap {
+        case (id, Right(patch)) => patch.asForwardUpdateRequest
+          .fold(error => SMono.just(ForwardSetUpdateFailure(id, error)),
+            validPath => update(MappingSource.fromUser(mailboxSession.getUser), getForwards(validPath, mailboxSession.getUser)))
+        case (id, Left(error)) => SMono.just(ForwardSetUpdateFailure(id, error))
+      }
+      .map(_.asForwardSetUpdateResult)
+      .foldWith[ForwardSetUpdateResults](ForwardSetUpdateResults.empty())(ForwardSetUpdateResults.merge)
+
+  private def getForwards(updateRequest: ForwardUpdateRequest, currentUser: Username): Seq[MailAddress] = {
+    if (updateRequest.localCopy.value.booleanValue()) {
+      updateRequest.forwards.map(_.value) :+ currentUser.asMailAddress()
+    } else {
+      updateRequest.forwards.map(_.value)
+    }
+  }
+
+  private def createResponse(invocation: Invocation,
+                             forwardSetRequest: ForwardSetRequest,
+                             updateResult: ForwardSetUpdateResults): Invocation = {
+    val response: ForwardSetResponse = ForwardSetResponse(
+      accountId = forwardSetRequest.accountId,
+      newState = UuidState.INSTANCE,
+      updated = Some(updateResult.updateSuccess).filter(_.nonEmpty),
+      notUpdated = Some(updateResult.updateFailures).filter(_.nonEmpty),
+      notCreated = validateNoCreate(forwardSetRequest),
+      notDestroyed = validateNoDestroy(forwardSetRequest))
+
+    Invocation(methodName,
+      Arguments(ForwardSerializer.serializeForwardSetResponse(response).as[JsObject]),
+      invocation.methodCallId)
+  }
+
+  private def update(mappingSource: MappingSource, forwardDestinations: Seq[MailAddress]): SMono[ForwardSetUpdateResult] =
+    retrieveMappings(mappingSource)
+      .map(currentForwards => (currentForwards.diff(forwardDestinations), forwardDestinations.diff(currentForwards)))
+      .flatMapMany {
+        case (deletedForwards, addedForwards) => deleteMappings(mappingSource, deletedForwards)
+          .thenMany(addMappings(mappingSource, addedForwards))
+      }.`then`()
+      .onErrorResume(error => SMono.just[ForwardSetUpdateResult](ForwardSetUpdateFailure("todo", error)))
```

Comment:
It is a bit weird to see a "todo" here. However, it seems not to impact things.

Reply:
Fixed
https://github.com/linagora/tmail-backend/pull/430/files